### PR TITLE
Init minio client on demand

### PR DESCRIPTION
**Pull request by gitlawr**
Issue: https://github.com/rancher/harvester/issues/33

So that server can run without minio service availability

### CODE REVIEW

**Comment by guangbochen** [comment]
@gitlawr can u please resolve the conflict

**Comment by gitlawr** [comment]
Updated

**Review by guangbochen — APPROVED** [review]
LGTM, the UI will need to address https://github.com/rancher/harvester/issues/50

**Comment by guangbochen** [comment]
@gitlawr will need to rebase this PR ,thanks

File: pkg/controller/master/image/controller.go [hunk 1]
```diff
@@ -97,7 +97,19 @@ func (h *Handler) OnImageChanged(key string, image *apisv1alpha1.Image) (*apisv1
 func (h *Handler) OnImageRemove(key string, image *apisv1alpha1.Image) (*apisv1alpha1.Image, error) {
 	logrus.Debugf("removing %s object in minio", image.Name)
 	cleanUpImport(image.Name, image.Status.AppliedURL)
-	return image, util.MinioClient.RemoveObject(util.BucketName, image.Name)
+	if err := removeImageFromStorage(image); err != nil {
+		// Log the error and do not block resource deletion.
+		logrus.Errorf("failed to remove image %q from storage: %v", image.Spec.DisplayName, err)
+	}
+	return image, nil
```
Comment (thxCode) [inline pkg/controller/master/image/controller.go]: Why cannot the user remove the image from storage? is it possible to return the error message to the frontend but not just log in the backend?

File: pkg/util/minio.go [hunk 2]
```diff
@@ -16,10 +16,18 @@ const (
...
+	minioClient *minio.Client
 )
 
-func InitMinio() (err error) {
+func GetMinioClient() (*minio.Client, error) {
+	var err error
+	if minioClient == nil {
+		err = initMinio()
+	}
+	return minioClient, err
```
Comment (thxCode) [inline pkg/util/minio.go]: It seems `GetMinioClient` function calls in three async places, but there is no synchronization primitive to protect `minioClient`. On the other hand, is it necessary to use a Single-Instance pattern in here?

**Comment by gitlawr** [comment]
@thxCode updated

**Review by thxCode — APPROVED** [review]
LGTM

**Pull request merged**